### ISSUE TITLE
UnusedPrivateClass - false positive for deserialization

### ISSUE THREAD

**Issue by imanushin** [detekt/detekt#2791]
## Expected Behaviour
No warnings are appeared 

## Observed Behavior
UnusedPrivateClass warning is appeared

## Steps to Reproduce
Code:
```kotlin
fun parse(content: String): List<String> {
      return jacksonObjectMapper().readValue<List<Response>>(content)
           .map { it.id }
}

private class Response(
     val id: String
)
```

## Context
This is a bit more complex case than #2158 . 
Such constructions are used for deserialization purpose, when you have private class which represents the external model.

## Your Environment
* Version of detekt used: 1.9.1
* Version of Gradle used (if applicable): 6.5.0
* Operating System and version: n/a
* Link to your project (if it's a public repository): private project

**Comment by cortinico** [comment]
I was able to reproduce. Here also a simpler scenario:
```kotlin
val elements = listOf(42).filterIsInstance<Set<Item>>()
private class Item
```
